etna/transforms/encoders/mean_encoder.py

```diff
@@ -0,0 +1,309 @@
+
+    category = "category"
+    global_mean = "global_mean"
+
+    @classmethod
+    def _missing_(cls, value):
+        raise NotImplementedError(
+            f"{value} is not a valid {cls.__name__}. Supported types: {', '.join([repr(m.value) for m in cls])}"
+        )
+
+
+class MeanEncoderTransform(IrreversibleTransform):
+    """
+    Makes encoding of categorical feature.
+
+    For timestamps that are before the last timestamp seen in ``fit`` transformations are made using the formula below:
+
+    .. math::
+       \\frac{TargetSum + RunningMean * Smoothing}{FeatureCount + Smoothing}
+
+    where
+
+        * TargetSum is the sum of target up to the current timestamp for the current category, not including the current timestamp
+        * RunningMean is target mean up to the current timestamp, not including the current timestamp
+        * FeatureCount is the number of categories with the same value as in the current timestamp, not including the current timestamp
+
+    For future timestamps:
+
+        * for known categories encoding are filled with global mean of target for these categories calculated during ``fit``
+        * for unknown categories encoding are filled with global mean of target in the whole dataset calculated during ``fit``
+
+    All types of NaN values are considering as one category.
+    """
+
+    idx = pd.IndexSlice
+
+    def __init__(
+        self,
+        in_column: str,
+        out_column: str,
+        mode: Union[EncoderMode, str] = "per-segment",
+        handle_missing: str = MissingMode.category,
+        smoothing: int = 1,
+    ):
+        """
+        Init MeanEncoderTransform.
+
+        Parameters
+        ----------
+        in_column:
+            categorical column to apply transform
+        out_column:
+            name of added column
+        mode:
+            mode to encode segments
+
+            * 'per-segment' - statistics are calculated across each segment individually
+
+            * 'macro' - statistics are calculated across all segments. In this mode transform can work with new segments that were not seen during ``fit``
+        handle_missing:
+            mode to handle missing values in ``in_column``
+
+            * 'category' - NaNs they are interpreted as a separate categorical feature
+
+            * 'global_mean' - NaNs are filled with the running mean
+        smoothing:
+            smoothing parameter
+        """
+        super().__init__(required_features=["target", in_column])
+        self.in_column = in_column
+        self.out_column = out_column
+        self.mode = EncoderMode(mode)
+        self.handle_missing = MissingMode(handle_missing)
+        self.smoothing = smoothing
+
+        self._global_means: Optional[Union[float, Dict[str, float]]] = None
+        self._global_means_category: Optional[Union[Dict[str, float], Dict[str, Dict[str, float]]]] = None
+        self._last_timestamp: Optional[Timestamp] = None
```

Comment:
It should have type: `Union[Timestamp, int, None`].

Reply:
timestamp can be None in TSDataset?

Reply:
Timestamp can be int.

Reply:
You could write `Optional[Union[Timestamp, int]]`, but it probably easier to write `Union[Timestamp, int, None]`.

Reply:
Fixed